src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -325,6 +331,120 @@ EnsureCitusTableCanBeCreated(Oid relationOid)
 }
 
 
+/*
+ * undistribute_table gets a distributed table name and
+ * udistributes the table.
+ */
+Datum
+undistribute_table(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid relationId = PG_GETARG_OID(0);
+
+	EnsureTableOwner(relationId);
+	Relation relation = relation_open(relationId, ExclusiveLock);
+
+	const char *relationName = get_rel_name(relationId);
+
+	Oid schemaId = get_rel_namespace(relationId);
+	const char *schemaName = get_namespace_name(schemaId);
+
+
+	if (TableReferencing(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because it has a foreign key.")));
+	}
+
+	if (TableReferenced(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because a foreign key references to it.")));
+	}
+
+	StringInfo query = makeStringInfo();
+
+	SPI_connect();
+
+	List *tableCommands = GetTableDDLEventsExceptCreation(relationId);
+	List *createTableCommands = GetTableCreationCommands(relationId, true);
+
+
+	StringInfo tempName = makeStringInfo();
+	CreateTemporaryTableName(&tempName, relationName);
+
+	char *createTableQuery = NULL;
+	foreach_ptr(createTableQuery, createTableCommands)
+	{
+		Node *parseTree = ParseTreeNode(createTableQuery);
+
+		if (nodeTag(parseTree) == T_CreateStmt)
+		{
+			((CreateStmt *) parseTree)->relation->relname = tempName->data;
+		}
+		else if (nodeTag(parseTree) == T_AlterOwnerStmt)
+		{
+			((AlterOwnerStmt *) parseTree)->relation->relname = tempName->data;
+		}
+		else if (nodeTag(parseTree) == T_AlterTableStmt)
+		{
+			((AlterTableStmt *) parseTree)->relation->relname = tempName->data;
+		}
+		else
+		{
+			continue;
+		}
+
+		CitusProcessUtility(parseTree, createTableQuery, PROCESS_UTILITY_TOPLEVEL, NULL,
+							None_Receiver, NULL);
+		break;
+	}
+
+	resetStringInfo(query);
+	appendStringInfo(query, "INSERT INTO %s SELECT * FROM %s",
+					 quote_qualified_identifier(schemaName, tempName->data),
+					 quote_qualified_identifier(schemaName, relationName));
+	SPI_execute(query->data, false, 0);
+
+	ObjectAddress undistributeTableAddres;
+
+	undistributeTableAddres.classId = RelationRelationId;
+	undistributeTableAddres.objectId = relationId;
+	undistributeTableAddres.objectSubId = 0;
+
+	DirectFunctionCall3(master_drop_all_shards,
+						ObjectIdGetDatum(relationId),
+						CStringGetTextDatum(relationName),
+						CStringGetTextDatum(schemaName));
+	DeletePartitionRow(relationId);
+
+	relation_close(relation, NoLock);
+
+	performDeletion(&undistributeTableAddres, 0, 0);
```

Comment:
Does this also drop the table from the workers when we are syncing metadata?

What happens to foreign keys at this point?